cirq/circuits/text_diagram_drawer.py

```diff
@@ -243,6 +260,73 @@ def render(self,
                          for row in multiline_grid
                          for sub_row in row).rstrip()
 
+    def _value_equality_values_(self):
+        return (self.entries,
+                self.vertical_lines, self.horizontal_lines,
+                self.vertical_padding, self.horizontal_padding)
+
+    def copy(self):
+        return self.__class__(
+                entries=self.entries,
+                vertical_lines=self.vertical_lines,
+                horizontal_lines=self.horizontal_lines,
+                vertical_padding=self.vertical_padding,
+                horizontal_padding=self.horizontal_padding)
+
+    def shifted(self, dx: int = 0, dy: int = 0) -> 'TextDiagramDrawer':
+        shifted_entries = {(x + dx, y + dy): t
+                for (x, y), t in self.entries.items()}
+        shifted_horizontal_lines = [
+                _HorizontalLine(y + dy, x1 + dx, x2 + dx, e)
+                for y, x1, x2, e in self.horizontal_lines]
+        shifted_vertical_lines = [
+                _VerticalLine(x + dx, y1 + dy, y2 + dy, e)
+                for x, y1, y2, e in self.vertical_lines]
+        shifted_horizontal_padding = {x + dx: padding
+                for x, padding in self.horizontal_padding.items()}
+        shifted_vertical_padding = {y + dy: padding
+                for y, padding in self.vertical_padding.items()}
+        return self.__class__(
+                entries=shifted_entries,
+                horizontal_lines=shifted_horizontal_lines,
+                vertical_lines=shifted_vertical_lines,
+                horizontal_padding=shifted_horizontal_padding,
+                vertical_padding=shifted_vertical_padding)
+
+    def __iadd__(self, other: 'TextDiagramDrawer') -> 'TextDiagramDrawer':
```

Comment:
I don't think that this is there is a natural unambiguous meaning for "adding" two diagrams. E.g. I would have expected addition to do horizontal concatenation, as it does for strings and lists in python. I think we should give this a name instead of an operator e.g. "merge".

Reply:
Changed to "update".